conans/client/graph/grapher.py

```diff
@@ -1,62 +1,138 @@
 import os
 import re
-
+from collections import OrderedDict
 from conans.client.graph.graph import BINARY_BUILD, BINARY_CACHE, BINARY_DOWNLOAD, BINARY_MISSING, \
     BINARY_UPDATE
 from conans.client.installer import build_id
 from conans.util.files import save
 from conans.errors import ConanException
 
 class ConanGrapher(object):
+
     def __init__(self, deps_graph):
         self._deps_graph = deps_graph
 
     def graph(self):
-        graph_lines = ['strict digraph {']
-        graph_lines.append(self._dot_configuration)
-        graph_lines.append('\n')
+        dot_graph = ['strict digraph {']
+        dot_graph.append(self._dot_configuration)
+        dot_graph.append('\n')
+
         # First, create the nodes
+        nodes = self._add_single_nodes_to_graph(dot_graph)
+
+        # Then, the adjacency matrix
+        self._add_adjacency_matrix(nodes, dot_graph)
+
+        dot_graph.append('}\n')
+
+        return ''.join(dot_graph)
+
+    def graph_file(self, output_filename):
+        save(output_filename, self.graph())
+
+    def _add_single_nodes_to_graph(self, dot_graph):
+        # Store list of build_requires nodes
+        build_time_nodes = self._deps_graph.build_time_nodes()
+
+        # Store the root node
+        root_node = self._deps_graph.root
+        if root_node.conanfile.name and root_node.conanfile.version:
+            root_id = "{}/{}".format(root_node.conanfile.name, root_node.conanfile.version)
+        elif root_node.conanfile.name:
+            root_id = root_node.conanfile.name
+        else:
+            root_id = root_node.conanfile.display_name
+
+        # Store nodes in ordered dict, for deterministic output to dot file
+        nodes = {}
+
+        # First, gather the nodes info
         for node in self._deps_graph.nodes:
-            node_id = node.conanfile.display_name
-            node_name = node.conanfile.name if node.conanfile.name else node.conanfile.display_name
-            node_version = node.conanfile.version if node.conanfile.version else ""
+            if node.conanfile.name and node.conanfile.version:
+                node_name = node.conanfile.name
+                node_version = node.conanfile.version
+                node_id = "{}/{}".format(node.conanfile.name, node.conanfile.version)
```

Comment:
It is not really needed to complicate the `node_id`. we can by default, set it to 

```
node_name = node.conanfile.display_name
```

To be compatible with the previous dot file definition. Otherwhise, the tests are really hard to be fixed (and information such user and channel should be obtained from the following structure

```
    "protobuf/3.5.2" [ fillcolor=azure, color=dodgerblue  label=<
     <table border="0" cellborder="0" cellspacing="0">
       <tr><td align="center"><b>protobuf</b></td></tr>
       <tr><td align="center"><font point-size="12">3.5.2</font></td></tr>
       <tr><td align="center"><i><font point-size="12">bincrafters/stable</font></i></td></tr>
     </table>>];
```
.